src/java.base/share/classes/jdk/internal/foreign/abi/NativeEntryPoint.java

```diff
@@ -80,13 +81,18 @@ public static NativeEntryPoint make(ABIDescriptor abi,
         });
     }
 
-    private static void checkType(MethodType methodType, boolean needsReturnBuffer, int savedValueMask) {
+    private static void checkType(MethodType methodType, boolean needsReturnBuffer, int savedValueMask,
+                                  boolean usingAddressPairs) {
         if (methodType.parameterType(0) != long.class) {
             throw new AssertionError("Address expected as first param: " + methodType);
         }
         int checkIdx = 1;
         if ((needsReturnBuffer && methodType.parameterType(checkIdx++) != long.class)
-            || (savedValueMask != 0 && methodType.parameterType(checkIdx) != long.class)) {
+            || (savedValueMask != 0 &&
```

Comment:
Maybe this routine is getting too complex and it would be better to split the checks (and add some comments) ? E.g. we need to check that if there's a return buffer, a certain low-level argument is `long` and, if there's need to capture state, we either have `long`, or an `Object,long` pair.

Reply:
Perhaps, even throwing different assertion error might help

Reply:
Yeah... I'll split this up a bit to make it more readable.